src/include/storage/garbage_collector.h

```diff
@@ -39,38 +44,177 @@ class GarbageCollector {
+   */
+  void CopyVarlen(UndoRecord *undo_record);
+
+  /**
+   * Given the start and the end of the compaction interval after first GC pass, do a second pass and create the
+   * compacted undo record
+   * @param start_record the undo record which marks the beginning of compacted interval
+   * @param end_record the undo record which marks the end of compacted interval
+   * @return the compacted undo record
+   */
+  UndoRecord *CreateUndoRecord(UndoRecord *start_record, UndoRecord *end_record);
+
+  /**
+   * Marks the end of two pass Interval GC
+   * @param interval_length_ptr the length of the compaction interval
+   */
+  void EndCompaction(uint32_t *interval_length_ptr);
+
+  /**
+   * Given the time stamp and table slot after first GC pass, create a undo record which can accommodate all the columns
+   * found in the combined delta in the first GC pass
+   * @param timestamp the timestamp which will be allotted to the compacted undo record
+   * @param slot the tuple slot corresponding to the tuple associated with the compacted undo record
+   * @param table the table corresponding to the tuple associated with the compacted undo record
+   * @return the blank undo record which is the placeholder for the compacted undo record
+   */
+  UndoRecord *InitializeUndoRecord(transaction::timestamp_t timestamp, TupleSlot slot, DataTable *table);
+
+  /**
+   * Given the compacted undo record to be linked to the version chain, link it
+   * @param start_record the undo record which marks the beginning of compacted interval and
+   * which will point to the compacted record
+   * @param curr_ptr the pointer marking the current position of interval GC in the version chain
+   * @param end_record the undo record which marks the end of compacted interval and compacted undo record will point
+   * to this record
+   * @param compacted_undo_record the compacted undo record to be linked to the version chain
+   */
+  void LinkCompactedUndoRecord(UndoRecord *start_record, UndoRecord **curr_ptr, UndoRecord *end_record,
+                               UndoRecord *compacted_undo_record);
+
   /**
    * Process the deallocate queue
    * @return number of txns (not UndoRecords) processed for debugging/testing
    */
   uint32_t ProcessDeallocateQueue();
 
+  /*
+   * Process deferred actions
+   */
+  void ProcessDeferredActions();
+
+  /**
+   * Given a version chain, perform interval gc on all versions except the head of the chain
+   * @param table data table
+   * @param slot tuple slot
+   * @param active_txns vector containing all active transactions
+   */
+  void ProcessTupleVersionChain(DataTable *table, TupleSlot slot, std::vector<transaction::timestamp_t> *active_txns);
+
+  /**
+   * Given the data table and the tuple slot, try unlinking the version chain head for that tuple slot
+   * @param table data table
+   * @param slot tuple slot
+   * @param active_txns list of currently running active transactions
```

Comment:
Style nit: list -> vector to be consistent with other function description and the actual container type